client/client_test.go

```diff
@@ -104,6 +107,100 @@ func TestNewEnvClient(t *testing.T) {
 	}
 }
 
+func TestClientWithContext(t *testing.T) {
+	caBytes, err := ioutil.ReadFile("testdata/ca.pem")
+	assert.Check(t, err)
+	certBytes, err := ioutil.ReadFile("testdata/cert.pem")
+	assert.Check(t, err)
+	keyBytes, err := ioutil.ReadFile("testdata/key.pem")
+	assert.Check(t, err)
+	storeDir, err := ioutil.TempDir("", "test-client-with-context")
+	assert.Check(t, err)
+	defer os.RemoveAll(storeDir)
+	type contextDef struct {
+		name          string
+		host          string
+		apiVersion    string
+		ca            []byte
+		cert          []byte
+		key           []byte
+		skipTLSVerify bool
+	}
+	contexts := []contextDef{
+		{
+			name: "all-default",
+		},
+		{
+			name: "host-only",
+			host: "https://notunixsocket",
+		},
+		{
+			name:       "apiversion-only",
+			apiVersion: "1.22",
+		},
+		{
+			name: "with-tls",
+			host: "https://notunixsocket",
+			ca:   caBytes,
+			key:  keyBytes,
+			cert: certBytes,
+		},
+		{
+			name:          "with-tls-skipverify",
+			host:          "https://notunixsocket",
+			ca:            caBytes,
+			key:           keyBytes,
+			cert:          certBytes,
+			skipTLSVerify: true,
+		},
+		{
+			name:          "with-skipverify-only",
+			host:          "https://notunixsocket",
+			skipTLSVerify: true,
+		},
+	}
+	s, err := contextstore.NewStore(storeDir)
+	assert.Check(t, err)
+	for _, c := range contexts {
+		t.Run(c.name, func(t *testing.T) {
+			err = context.SetDockerEndpoint(s, c.name, c.host, c.apiVersion, c.ca, c.cert, c.key, c.skipTLSVerify)
```

Comment:
I wonder if it would be better to do all these first (with a new for loop before this one), reason is this will better test the functionality of looking up one among many options, whereas with it here the first test only has one option in the store + it is always looking up the last thing added.